sqlliveness: add CachedReader to avoid synchronous lookups

ajwerner opened the pull request:
Sometimes you want to know the liveness status of a session in a best-effort
way that avoids any RPCs. The liveness status reported by the Reader is already
best-effort (a session may be removed in the meantime). This minor change
allows callers to opt into the more optimistic caching semantics.

Release note: None

ajwerner committed:
sql/sqlliveness: add CachedReader to avoid synchronous lookups

Sometimes you want to know the liveness status of a session in a best-effort
way that avoids any RPCs. The liveness status reported by the Reader is already
best-effort (a session may be removed in the meantime). This minor change
allows callers to opt into the more optimistic caching semantics.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67075)
<!-- Reviewable:end -->

ajwerner committed:
sql/sqlliveness/slstorage: use new context in singleflight

This fixes a bug whereby the context of the caller which launched the
singleflight might have resulted in an error being returned to the other
goroutines. This also reveals another, more serious bug: the wrong error
was being returned from the singleflight.

Release note (bug fix): Fixed a bug where job leases might be revoked due to
a transient network error.

ajwerner committed:
sql/sqlliveness/slstorage: allow reader to be canceled

Release note: None

rimadeodhar reviewed (approved):
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67075)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @dhartunian)

___
*[pkg/sql/sqlliveness/slstorage/slstorage.go, line 171 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67075#-MdSp0PC2ZFVQMAJNfxj:-MdSp0PC2ZFVQMAJNfxk:bi7g6fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/57ca751d2be2871f120b1950675c90ff131b0dfe/pkg/sql/sqlliveness/slstorage/slstorage.go#L171)):*
> ```Go
> // session definitely is not alive.
> func (s *Storage) IsAlive(ctx context.Context, sid sqlliveness.SessionID) (alive bool, err error) {
> 	const synchronousLookup = false
> ```

Nit: Shouldn't this be asynchronousLookup?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR!

bors r=rimadeodhar

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67075#-:-Md_P3Tq2EFp_JxwtgDV:b-368ct5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @rimadeodhar)

___
*[pkg/sql/sqlliveness/slstorage/slstorage.go, line 171 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67075#-MdSp0PC2ZFVQMAJNfxj:-MdUigwiAeLsAu65ukK1:bayhxw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/57ca751d2be2871f120b1950675c90ff131b0dfe/pkg/sql/sqlliveness/slstorage/slstorage.go#L171)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Shouldn't this be asynchronousLookup?
</blockquote></details>

The idea is that `false` here means `synchronousLookup`. Maybe that's confusing. I've reworked it to use an enum for maximum clarity (almost definitely overkill). 


<!-- Sent from Reviewable.io -->
